homeassistant/components/mqtt/__init__.py

```diff
@@ -299,6 +302,36 @@ def publish_template(hass: HomeAssistantType, topic, payload_template,
     hass.services.call(DOMAIN, SERVICE_PUBLISH, data)
 
 
+def wrap_callback(func):
+    """Decorate an MQTT message callback to catch and log exceptions."""
+    def log_exception(topic, payload):
+        module_name = inspect.getmodule(inspect.trace()[1][0]).__name__
+        logging.getLogger(module_name).error(
+            "Exception in %s when handling msg on '%s': '%s'\n%s",
+            func.__name__, topic, payload, traceback.format_exc(limit=-1))
+
+    wrapper_func = None
+    if asyncio.iscoroutinefunction(func):
```

Comment:
Why not change this to: 
```

        @wraps(func)
        def wrapper(topic, payload, qos):
            """Catch and log exception."""
            try:
                if asyncio.iscoroutinefunction(func):
                    await func(topic, payload, qos)
                else:
                    func(topic, payload, qos)
            except Exception:  # pylint: disable=broad-except
                log_exception(topic, payload)
        wrapper_func = wrapper
```

Reply:
I like this, note that we should also handle threaded functions (non-callback).

Reply:
@Danielhiversen If changed, the caller doesn't know how to call the wrapper.

Reply:
Ahh, I see

Reply:
> I like this, note that we should also handle threaded functions (non-callback).

@balloob Meaning this PR is OK, or meaning to change according to @Danielhiversen's comment?

Reply:
Daniels comment.

Reply:
Actually, I think your current implementation is fine.